crates/cli/src/commands/format.rs

```diff
@@ -0,0 +1,176 @@
+    }
+
+    Ok(())
+}
+
+fn format_yaml_file(file_content: &str) -> Result<String> {
+    // deserializing manually and not using `SerializedGritConfig` because
+    // i don't want to remove any fields that `SerializedGritConfig` don't have such as 'version'
+    let mut config: serde_yaml::Value =
+        serde_yaml::from_str(file_content).with_context(|| "couldn't parse yaml file")?;
+    let patterns = config
+        .get_mut("patterns")
+        .ok_or_else(|| anyhow!("couldn't find patterns in yaml file"))?
+        .as_sequence_mut()
+        .ok_or_else(|| anyhow!("patterns in yaml file are not sequence"))?;
+    for pattern in patterns {
+        let Some(body) = pattern.get_mut("body") else {
+            continue;
+        };
+        if let serde_yaml::Value::String(body_str) = body {
+            *body_str = format_grit_code(body_str)?;
+            // extra new line at end of grit body looks more readable
+            body_str.push('\n');
+        }
+    }
+    Ok(serde_yaml::to_string(&config)?)
+}
+
+fn format_pattern_as_hunk_changes(pattern: &ResolvedGritDefinition) -> Result<HunkChange> {
+    let formatted_grit_code = format_grit_code(&pattern.body)?;
+    let body_range = pattern
+        .config
+        .range
+        .ok_or_else(|| anyhow!("pattern doesn't have config range"))?;
+    Ok(HunkChange {
+        starting_byte: body_range.start_byte as usize,
+        ending_byte: body_range.end_byte as usize,
+        new_content: formatted_grit_code,
+    })
+}
+
+/// format grit code using `biome`
+fn format_grit_code(source: &str) -> Result<String> {
+    let parsed = biome_grit_parser::parse_grit(source);
+    ensure!(
+        parsed.diagnostics().is_empty(),
+        "biome couldn't parse: {}",
+        parsed
+            .diagnostics()
+            .iter()
+            .map(|diag| diag.message.to_string())
+            .collect::<Vec<_>>()
+            .join("\n")
+    );
+
+    let options = GritFormatOptions::default();
+    let doc = biome_grit_formatter::format_node(options, &parsed.syntax())
+        .with_context(|| "biome couldn't format")?;
+    Ok(doc.print()?.into_code())
+}
+
+/// Represent a hunk of text that needs to be changed
+#[derive(Debug)]
+struct HunkChange {
+    starting_byte: usize,
+    ending_byte: usize,
+    new_content: String,
+}
+
+/// returns a new string that applies hunk changes
+fn apply_hunk_changes(input: &str, mut hunks: Vec<HunkChange>) -> String {
+    if hunks.is_empty() {
+        return input.to_string();
+    }
+    hunks.sort_by_key(|hunk| hunk.starting_byte);
+    let mut buffer = String::new();
+    let mut last_ending_byte = 0;
+    for (index, hunk) in hunks.iter().enumerate() {
+        buffer.push_str(&input[last_ending_byte..hunk.starting_byte]);
+        buffer.push_str(&hunk.new_content);
+        last_ending_byte = hunk.ending_byte;
+
+        if index == hunks.len() - 1 {
+            buffer.push_str(&input[last_ending_byte..]);
+        }
+    }
+    buffer
+}
```

Comment:
this could easily get removed if markdown could only have one grit pattern (and already all examples that i could find in `stdlib` only has one pattern per markdown), but i remember that when parsing markdown it could technically handle multiple patterns, so i just kept it

Reply:
Some use cases do have multiple patterns (usually for internal use) so we can't assume it.